{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Guide to software/module/package installation\n",
    "This section records the installation I have done in the servers I used thus far"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#R packages\n",
    "#Trying to plot a venn diagram, which needs library(\"VennDiagram\") and library(\"extrafont\")\n",
    "#Run default R.\n",
    "[14:03:02] jlee@zeus: /group/pawsey0149/groupEnv/ivec/lib $ which R\n",
    "/pawsey/sles11sp4/apps/gcc/4.8.5/r/3.2.2/bin/R\n",
    "#Install, choose mirror, make sure the dir is specified as \"/group/pawsey0149/groupEnv/ivec/lib/R\"\n",
    "> install.packages(\"VennDiagram\")\n",
    "#In the case of error as following\n",
    "Warning: unable to access index for repository https://mirrors.tuna.tsinghua.edu.cn/CRAN/src/contrib\n",
    "Warning message:\n",
    "package ‘VennDiagram’ is not available (for R version 3.2.2) \n",
    "#do a curl to find out whether that link is valid\n",
    "[14:01:07] jlee@zeus: /group/pawsey0149/groupEnv/ivec/lib $ curl https://mirrors.tuna.tsinghua.edu.cn/CRAN/src/contrib\n",
    "<html>\n",
    "<head><title>301 Moved Permanently</title></head>\n",
    "<body bgcolor=\"white\">\n",
    "<center><h1>301 Moved Permanently</h1></center>\n",
    "<hr><center>nginx/1.10.1</center>\n",
    "</body>\n",
    "</html>\n",
    "#Apparently it was moved and redirected\n",
    "#So, add slash for the redirection, and remove the \"s\" in https\n",
    "> install.packages(\"VennDiagram\", contriburl=\"http://mirrors.tuna.tsinghua.edu.cn/CRAN/src/contrib/\")\n",
    "#Success!\n",
    "\n",
    "#Somtimes there is this error about \"slam\"\n",
    ">  install.packages(\"wordcloud\", contriburl=\"http://mirrors.tuna.tsinghua.edu.cn/CRAN/src/contrib/\")\n",
    "Installing package into ‘/group/pawsey0149/groupEnv/ivec/lib/R’\n",
    "(as ‘lib’ is unspecified)\n",
    "Warning: dependency ‘slam’ is not available\n",
    "trying URL 'http://mirrors.tuna.tsinghua.edu.cn/CRAN/src/contrib//wordcloud_2.5.tar.gz'\n",
    "Content type 'application/octet-stream' length 40861 bytes (39 KB)\n",
    "==================================================\n",
    "downloaded 39 KB\n",
    "\n",
    "ERROR: dependency ‘slam’ is not available for package ‘wordcloud’\n",
    "* removing ‘/group/pawsey0149/groupEnv/ivec/lib/R/wordcloud’\n",
    "\n",
    "The downloaded source packages are in\n",
    "\t‘/tmp/Rtmp44v8YJ/downloaded_packages’\n",
    "Warning message:\n",
    "In install.packages(\"wordcloud\", contriburl = \"http://mirrors.tuna.tsinghua.edu.cn/CRAN/src/contrib/\") :\n",
    "  installation of package ‘wordcloud’ had non-zero exit status\n",
    "\n",
    "#solution\n",
    "sudo apt-get install r-cran-slam\n",
    "#then install again\n",
    "\n",
    "\n",
    "b) python in Pawsey - Zeus\n",
    "\n",
    "#first, download the actual archive of the package you want to install\n",
    "wget http://some_url/package.tar.gz\n",
    "#extract\n",
    "tar xzvf package.tar.gz\n",
    "cd package\n",
    "python setup.py install --prefix=$MYGROUP/../groupEnv/ivec/lib/python\n",
    "# in case there are problems, check PYTHONPATH variable\n",
    "# Then change the file and folder permission\n",
    "y82_addGroupPermission $MYGROUP/../groupEnv/ivec/lib/python\n",
    "\n",
    "Always test whether it installed sucessfully, the laziest way is to do\n",
    "python -c \"import pysam\" # or whatever your name is\n",
    "BUT don't do that inside the directory of the package you used for installation,\n",
    "as Python will import the package from there instead from where you just installed to,\n",
    "so test the package from a different working directory.\n",
    "\n",
    "c) Update ldd version\n",
    "Some programs need 2.14 glibc version to compile. Example of error messages:\n",
    "ImportError: /lib64/libc.so.6: version `GLIBC_2.14' not found\n",
    "\n",
    "1. Do a check for your version\n",
    "[10:08:47] jlee@zeus: /scratch/pawsey0149/jlee/halophila/2_geneloss/ZMA_vs_ATH $ ldd --version\n",
    "ldd (GNU libc) 2.11.3\n",
    "Copyright (C) 2009 Free Software Foundation, Inc.\n",
    "This is free software; see the source for copying conditions.  There is NO\n",
    "warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n",
    "Written by Roland McGrath and Ulrich Drepper.\n",
    "\n",
    "2. Trying on raijin, locally install a glibc version\n",
    "cd ~\n",
    "wget https://ftp.gnu.org/gnu/glibc/glibc-2.19.tar.gz\n",
    "tar -xzf glibc-2.19.tar.gz\n",
    "cd glibc-2.19\n",
    "mkdir build\n",
    "cd build\n",
    "../configure  --prefix=/home/571/jxl571\n",
    "make\n",
    "make install\n",
    "[NEVER USE CONDA TO INSTALL NEWER GLIBC! EVERYTHING WILL BE SEG FAULT DUE TO MISMATCH OF LD.SO AND LIBC.SO, ETC]\n",
    "\n",
    "3. point the path to that version?\n",
    "add the path with ldd (bin) to PATH\n",
    "also add another path to LD_LIBRARY\n",
    "\n",
    "d) R in zeus - installation by source package\n",
    "1. first download the package to some path\n",
    "2. then run install.packages by specifying the path, easy\n",
    "\n",
    "> install.packages(\"/home/jlee/downloads/EMMIX_1.0.1.tar\", repos= NULL, type=\"source\")\n",
    "Installing package into ‘/group/pawsey0149/groupEnv/ivec/lib/R’\n",
    "(as ‘lib’ is unspecified)\n",
    "* installing *binary* package ‘EMMIX’ ...\n",
    "* DONE (EMMIX)\n",
    ">library(EMMIX)\n",
    "\n",
    "\n",
    "4. Cray compiler to make Canu\n",
    "Error:\n",
    "make\n",
    "Invalid characters after option '-M' in command line item '-MD'.\n",
    "SOlution:\n",
    "module swap PrgEnv-cray/5.2.82 PrgEnv-gnu/5.2.82\n",
    "delete canu folder\n",
    "git clone again\n",
    "make again\n",
    "SUCCESS!\n",
    "\n",
    "PrgEnv-cray\n",
    "The cce (Cray Compiling Environment) module should be changed. Use, e.g.,:\n",
    "magnus-1:~> module swap cce cce/8.2.0\n",
    "PrgEnv-intel\n",
    "The intel compiler module should be changed, e.g.,\n",
    "magnus-1:~> module swap intel intel/13.1.3.192\n",
    "PrgEnv-gnu\n",
    "The gcc module should be changed, e.g.,\n",
    "magnus-1:~> module swap gcc gcc/4.8.2\n",
    "\n",
    "\n",
    "for bashrc mistakes\n",
    "\n",
    "I added a path export line in my ~/.bashrc yesterday and it is possible that there is a mistake, as I have problems \n",
    "with logging in after that. Since bashrc is immediately sourced once I login, I get logged out almost immediately with this message:\n",
    "Connection to raijin5.nci.org.au closed.\n",
    "Rename like this:\n",
    "ssh -v jxl571@raijin5.nci.org.au '/bin/sh -c \"cp .bashrc .bashrc_bk; rm .bashrc\"'\n",
    "or\n",
    "Next time, you can ssh in without executing your profile:\n",
    "ssh –t  00090575@130.95.76.250 bash --norc --noprofile\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## s400"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#programs that are already installed\n",
    "/opt/blast/ncbi-blast-2.2.26+/bin\n",
    "/sw/Fastx/bin\n",
    "/sw/cufflinks-1.3.0.Linux_x86_64\n",
    "/sw/R-3.3.1/bin\n",
    "/sw400/bin/samtools\n",
    "\n",
    "#anaconda\n",
    "wget https://repo.anaconda.com/archive/Anaconda2-5.2.0-Linux-x86_64.sh\n",
    "sh Anaconda2-5.2.0-Linux-x86_64.sh\n",
    "\n",
    "#mummer-3.23\n",
    "conda install -c bioconda mummer \n",
    "\n",
    "#bowtie2\n",
    "conda install -c bioconda bowtie2\n",
    "\n",
    "#allmaps\n",
    "pip install biopython numpy deap networkx matplotlib jcvi\n",
    "wget --no-check-certificate https://www.dropbox.com/s/onsjieazu0uytgk/ALLMAPS-install.sh\n",
    "sh ALLMAPS-install.sh\n",
    "wget http://hgdownload.cse.ucsc.edu/admin/exe/linux.x86_64/liftOver\n",
    "chmod a+x liftOver\n",
    "\n",
    "#minimap and minidot\n",
    "git clone https://github.com/lh3/minimap && (cd minimap && make)\n",
    "git clone https://github.com/lh3/miniasm && (cd miniasm && make)\n",
    "git clone https://github.com/thackl/minidot.git\n",
    "#then in R\n",
    "install.packages(\"argparse\", contriburl=\"http://mirrors.tuna.tsinghua.edu.cn/CRAN/src/contrib/\")\n",
    "install.packages(\"ggplot2\")\n",
    "#exporting in myprofile\n",
    "export PATH=/home/s400/home00/pz1/lee-j/programs/minidot/bin:$PATH #minidot calls R script which is one folder outside of the bin within minidot, so\n",
    "\n",
    "#exonerate\n",
    "conda install -c bioconda exonerate\n",
    "\n",
    "#get newer R\n",
    "conda install -c r r #only get 3.4.3\n",
    "#get seletion tools package\n",
    "wget http://population-genetics.uni-giessen.de/~software/SelectionTools_18.2_R_x86_64-pc-linux-gnu.tar.gz\n",
    "#execute this way \n",
    "~/anaconda2/bin/R\n",
    "#then the package\n",
    "install.packages('/home/s400/home00/pz1/lee-j/programs/SelectionTools_18.2_R_x86_64-pc-linux-gnu.tar.gz', repos=NULL, type=\"source\")\n",
    "* installing *binary* package ‘SelectionTools’ ...\n",
    "* DONE (SelectionTools)\n",
    "#then load\n",
    "> library(SelectionTools)\n",
    "Error: package or namespace load failed for ‘SelectionTools’ in dyn.load(file, DLLpath = DLLpath, ...):\n",
    " unable to load shared object '/home/s400/home00/pz1/lee-j/anaconda2/lib/R/library/SelectionTools/libs/SelectionTools.so':\n",
    "  /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /home/s400/home00/pz1/lee-j/anaconda2/lib/R/library/SelectionTools/libs/SelectionTools.so)\n",
    "#problem!\n",
    "#R version 3.3.1\n",
    "conda install -c r r=3.3.1\n",
    "> install.packages('/home/s400/home00/pz1/lee-j/programs/SelectionTools_18.2_R_x86_64-pc-linux-gnu.tar.gz', repos=NUL* installing *binary* package ‘SelectionTools’ ...\n",
    "* DONE (SelectionTools)\n",
    "> st.read.marker.data\n",
    "Error: object 'st.read.marker.data' not found\n",
    "> library(SelectionTools)\n",
    "Error in dyn.load(file, DLLpath = DLLpath, ...) : \n",
    "  unable to load shared object '/home/s400/home00/pz1/lee-j/anaconda2/lib/R/library/SelectionTools/libs/SelectionTools.so':\n",
    "  /lib64/libc.so.6: version `GLIBC_2.14' not found (required by /home/s400/home00/pz1/lee-j/anaconda2/lib/R/library/SelectionTools/libs/SelectionTools.so)\n",
    "In addition: Warning message:\n",
    "package ‘SelectionTools’ was built under R version 3.3.2 \n",
    "Error: package or namespace load failed for ‘SelectionTools’\n",
    "#remove\n",
    "conda remove  r=3.3.1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SKYLLA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#bismark\n",
    "git clone https://github.com/FelixKrueger/Bismark.git\n",
    "\n",
    "#miniconda\n",
    "wget https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh\n",
    "sh Miniconda2-latest-Linux-x86_64.sh\n",
    "\n",
    "#samtools\n",
    "conda install -c bioconda samtools\n",
    "\n",
    "#chromosomer\n",
    "pip install chromosomer\n",
    "\n",
    "#missing perl module, starts with cpanm installation\n",
    "#install local::lib\n",
    "wget https://cpan.metacpan.org/authors/id/H/HA/HAARG/local-lib-2.000024.tar.gz\n",
    "tar -xzf local-lib-2.000024.tar.gz\n",
    "cd local-lib-2.00002\n",
    "perl Makefile.PL --bootstrap\n",
    "make test && make install\n",
    "echo 'eval \"$(perl -I$HOME/perl5/lib/perl5 -Mlocal::lib)\"' >>~/.bashrc\n",
    "#install cpanminus\n",
    "curl -L http://cpanmin.us | perl - -l ~/perl5 App::cpanminus local::lib\n",
    "#install whatever you want! hooray!\n",
    "cpanm GD::Graph::lines #problem, when submitted to queue, module still not found even installed, meaning some exporting is not done in cluster, have to figure this out....\n",
    "\n",
    "cpanm Bio::SeqIO          \n",
    "\n",
    "#install own version of newer glibc ##doesn't work, gave up.\n",
    "wget \n",
    "tar -xzf glibc-2.19.tar.gz\n",
    "cd glibc-2.19\n",
    "mkdir build\n",
    "cd build\n",
    " ../configure --prefix=/home/gh2743/programs/glibc-install\n",
    "make\n",
    "make install\n",
    "export LD_LIBRARY_PATH=/home/gh2743/programs/glibc-install/lib:$LD_LIBRARY_PATH\n",
    "/lib64/ld-linux-x86-64.so.2 --library-path /lib64 /bin/ls\n",
    "#this is default\n",
    "export LD_LIBRARY_PATH=/cm/shared/apps/gcc/4.8.1/lib:/cm/shared/apps/gcc/4.8.1/lib64:/cm/shared/apps/openmpi/gcc/64/1.6.5/lib\n",
    "\n",
    "\n",
    "        \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lummerland"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#putty is too ugly so I downloaded some themes\n",
    "https://github.com/jacktrocinski/pretty-putty\n",
    "    \n",
    "#example of array jobs submission in lummerland\n",
    "vi 2_bismark.sh\n",
    "#!/bin/bash\n",
    "\n",
    "#$ -S /bin/bash\n",
    "#$ -N bismark_step_2        #jobname\n",
    "#$ -t 1:137             #array jobs from 1 to 137, variable is $SGE_TASK_ID, does not accept 0\n",
    "                        #other variables are $SGE_TASK_FIRST, $SGE_TASK_LAST, $SGE_TASK_STEPSIZE\n",
    "#$ -tc 137              #how many max jobs run at parallel\n",
    "#$ -pe multislot 8      #request 1 core for each job, this does not limit the number of cores used by the tool\n",
    "#$ -cwd                 #use current dir as working dir\n",
    "# $ -o $SGE_TASK_ID.out         #output file\n",
    "# $ -e $SGE_TASK_ID.err         #error file\n",
    "# $ -wd /fhgfs/gh2743/BreedPath_06062018/workingdir/bismark/NRGene1.2             #working directory during submission\n",
    "#$ -j n            #stderr -> stdout redirection\n",
    "#$ -q all.q         #job-queue, all.q, interactive.q, idle.q, large.q, submit.q, imediate.q\n",
    "# $ -l virtual_free=32GB   #maximum requested memory per core\n",
    "\n",
    "#export OMP_NUM_THREADS=8\n",
    "#export OMP_THREAD_LIMIT=8\n",
    "\n",
    "date #log start time\n",
    "\n",
    "WORKING_DIR=\"/nfs/agipz/jlee/BreedPath_06062018/1_bisulphite\"\n",
    "SAMTOOLS=\"/ceph/biotools/bin\" #dont include executable\n",
    "BOWTIE=\"/ceph/biotools/bin\"\n",
    "\n",
    "cd $WORKING_DIR\n",
    "\n",
    "####bismark [options] <genome_folder> {-1 <mates1> -2 <mates2> | <singles>}\n",
    "\n",
    "#do this if your file cannot be renamed to SGE_TASK_ID\n",
    "#for f in  *_1.fq.gz  ; do echo $f; done > fq1.list\n",
    "#for f in  *_2.fq.gz  ; do echo $f; done > fq2.list\n",
    "READ1=$(head -n $SGE_TASK_ID fq1.list | tail -n 1)\n",
    "READ2=$(head -n $SGE_TASK_ID fq2.list | tail -n 1)\n",
    "\n",
    "cmd=\"/homes/jlee/programs/bin/bismark genome_input_for_bismark -o /nfs/agipz/jlee/BreedPath_06062018/1_bisulphite --parallel 8 --path_to_bowtie $BOWTIE --samtools_path $SAMTOOLS -1 $READ1 -2 $READ2\"\n",
    "\n",
    "$cmd\n",
    "\n",
    "echo $cmd\n",
    "\n",
    "echo done\n",
    "date\n",
    "\n",
    "exit\n",
    "\n",
    "#install runBNG\n",
    "git clone https://github.com/AppliedBioinformatics/runBNG.git\n",
    "#also need the bionano scripts\n",
    "wget http://www.bnxinstall.com/RefAlignerAssembler/Linux/AVX/tools_v5122_AVX.tar.gz\n",
    "wget http://www.bnxinstall.com/Scripts/v5134/scripts_v5134.tar.gz \n",
    "tar -xzf *tar.gz #test run if Assembler and RefAligner are executatble\n",
    "chmod +x runBNG\n",
    "#bionano scripts are only running on perl 5.10-5.18, which are SO OLD. I have 5.22 so I have to reinstall an older version\n",
    "#install perl 5.16\n",
    "wget https://www.cpan.org/src/5.0/perl-5.16.0.tar.gz\n",
    "tar -xzf perl-5.16.0\n",
    "cd perl-5.16.0\n",
    "sh Configure -Dprefix=/homes/jlee/programs/perl5.16\n",
    "make\n",
    "make test\n",
    "make install PERLNAME=perl5.16\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rstudio"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#downloaded the Selection Tools package from Frisch's group and install:\n",
    "http://population-genetics.uni-giessen.de/~software/\n",
    "install.packages('C://Users/gh2743/Downloads/SelectionTools_18.3.zip', repos=NULL, type=\"source\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# unix/awk/shell/etc cheatsheet\n",
    "This section records some commonly used hack that I tend to forget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Assign variable in shell into awk:\n",
    "for f in chrlg1_ExR53_weird_markers_type1.txt; do fgrep -w -f $f 60k_SNP_probe_seq_ids_positiononBNS14013_conversion_length50_gap0_uniq_pid100.txt; chro=`echo $f | awk -F \"_\" '{print $1}'` ; done | cut -f1 |  awk -v chr=\"$chro\" -F \"\\t\" '{print $1\"\\t\"chr}'\n",
    "#another example\n",
    "for f in MSLfragScaff_scaffold_109_fragScaff_scaffold_73.paf\n",
    "do\n",
    "query=`echo $f | sed 's/MSL//' | awk -F \"_\" '{print $1\"_\"$2\"_\"$3}'`\n",
    "awk  -F \"\\t\" '$1!=$6 ' $f | awk -F \"\\t\" -v name=\"$query\" '$1 ~ name '\n",
    "\n",
    "#Check last touched time of file\n",
    "find -type f -printf '%T+ %p\\n' | sort >time.txt\n",
    "\n",
    "#to save jupyter notebook as html\n",
    "ipython nbconvert notebook.ipynb\n",
    "\n",
    "#To pscp from server to windows pc, start command prompt on windows, then:\n",
    "Z:\\>pscp lee-j@fb09-pg-s400.agrar.uni-giessen.de:/clc03/48b_BreedPath_Birgit_Samans_02_2018/BreedPath/Proposal_reports/BreedPath_Vorhabenbeschreibung_gesamt.pdf C:\\Users\\gh2743\\Documents\\BreedPath_Vorhabenbeschreibung_gesamt.pdf\n",
    "Z:\\>pscp C:\\Users\\gh2743\\Documents\\ExR53_23k_pp_pa_1mMorgan.csv lee-j@fb09-pg-s400.agrar.uni-giessen.de:/home/s400/data01/lee-j/BreedPath_06062018/data/genetic_map \n",
    "#make sure files are not opened/are saved else it will say file not found\n",
    "\n",
    "# To run vncserver for proper size\n",
    "vncserver  -geometry 1920x1080 -depth 24\n",
    "\n",
    "#my profile for s400\n",
    "###### Alias\n",
    "alias ls='ls --color=always'\n",
    "alias ll='ls -l'\n",
    "alias cp='cp -i'\n",
    "alias rm='rm -i'\n",
    "alias mycp='rsync -avzh --progress'\n",
    "alias mac2unix=\"perl -pi -e 's/\\r/\\n/g' \"\n",
    "alias ws1=\"cd /home/s400/data01/lee-j\"\n",
    "alias ws2=\"cd /home/s400/data02/lee-j\"\n",
    "##### Prompt\n",
    "PS1=\"[\\t] \\u@\\h: \\w $ \"\n",
    "\n",
    "#then source it in .profile\n",
    "\n",
    "#find out the encoding of a file\n",
    "file -i BreedPath-Genotypisierung.csv \n",
    "BreedPath-Genotypisierung.csv: text/plain; charset=iso-8859-1\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Potentially useful tools\n",
    "This section records tools that look like I would use in the future\n",
    "\n",
    "TreeView 3 - Preferred heatmap viewer due to it's ability to control color template \n",
    "\n",
    "Java Tree View - Very flexible cluster/heatmap viewer\n",
    "\n",
    "Open Access\n",
    "\n",
    "LTR_retriever: A Highly Accurate and Sensitive Program for Identification of Long Terminal Repeat Retrotransposons\n",
    "\n",
    "METHimpute: imputation-guided construction of complete methylomes from WGBS data\n",
    "\n",
    "Draw LD stuff http://mg2c.iask.in/mg2c_v2.0/\n",
    "\n",
    "Bandage: visualization of de novo assemblies\n",
    "\n",
    "Kablammo - blast visualizer\n",
    "\n",
    "OmicCircos"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
